examples/06 Code Generation.ipynb

```diff
@@ -0,0 +1,430 @@
+    "        multiplicity = (f\"{prop.type.name}\", f\"list[{prop.type.name}]\")[\n",
+    "            prop.max_card.value > 1.0\n",
+    "        ]\n",
+    "        text.append(f\"\\t{prop.name}: {multiplicity}\")\n",
+    "    text = \"\\n\".join(text)\n",
+    "\n",
+    "    return text"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "# file: Trajectory.py \n",
+      "class Example:\n",
+      "\ttest: str\n",
+      "\n",
+      "\n",
+      "class Waypoint:\n",
+      "\tlat: float\n",
+      "\tlon: float\n",
+      "\talt: float\n",
+      "\texamples: list[Example]\n",
+      "\n",
+      "\n",
+      "class Trajectory:\n",
+      "\twaypoints: list[Waypoint]\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "trajectory = data_pkg.classes.by_name(\"Trajectory\")\n",
+    "text = class_to_python(trajectory)\n",
+    "filename = f\"{trajectory.name}.py\"\n",
+    "with open(filename, \"w\") as file:\n",
+    "    file.write(text)\n",
+    "print(f\"# file: {filename} \\n{text}\\n\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Interface for Protocol Buffers (Protobuf) \n",
+    "\n",
+    "Protobuf Message descriptions are stored in .proto files where class defintion starts with _message_ and each property of the class is definded by at least three parts: the data type, name and their order number. Classes can be also nested in classes. An example is shown In the following:\n",
+    "\n",
+    "### _class.proto_\n",
+    "\n",
+    "syntax = \"proto2\";\n",
+    "\n",
+    "message class1 {\n",
+    "\n",
+    ">    datatype class1_name1 = 1;\n",
+    "\n",
+    ">    dataype class1_name2 = 2;\n",
+    "\n",
+    ">    message class2 {\n",
+    "\n",
+    "> >  datatype class2_name1 = 1;\n",
+    "\n",
+    "> >  datatype class2_name2 = 2;\n",
+    "\n",
+    "> >  datatype class2_name3 = 3;\n",
+    "\n",
+    " >   }\n",
+    "\n",
+    ">    repeated class2 class_name = 3;\n",
+    "    \n",
+    "}\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def class_to_proto(cls, current_classes=[], indent=\"\\t\"):\n",
+    "    if not current_classes:\n",
+    "        text = ['syntax = \"proto3\";\\n']\n",
+    "    text = [f\"{indent[:-1]}message  {cls.name} {{\"]\n",
+    "    current_classes.append([cls])\n",
+    "\n",
+    "    for counter, prop in enumerate(cls.properties):\n",
+    "        multiplicity = (\"\", \"[]\")[prop.max_card.value > 1.0]\n",
+    "        if (\n",
+    "            isinstance(\n",
+    "                prop.type, capellambse.model.crosslayer.information.Class\n",
+    "            )\n",
+    "            and prop.type not in current_classes\n",
+    "        ):\n",
+    "            current_classes.append([prop.type])\n",
```

Comment:
You're checking whether `prop.type` is in current_classes, and if it's not, you're adding `[prop.type]` (note the brackets). Same thing where `[cls]` is added to the list earlier on. This means the containment check will still fail the next time we arrive here with the same argument, as it's not `prop.type` that's in `current_classes`, but rather a one-element list containing it. Try the following:

```python
>>> proptype = object()
>>> classes = []
>>> classes.append([proptype])
>>> proptype in classes
False
```